File: newton-method/newton_method.py
import math
import sympy as sp
sp.init_printing(use_unicode=True)


EPSILON = 1e-12
MAX_ITERS = 100
x = sp.Symbol('x')


def newton_method(f: sp.Expr, p0: float, max_iters: int = MAX_ITERS, tol: float = EPSILON) -> float | None:
    f_diff = sp.diff(f, x)
    f_lambda = sp.lambdify(x, f, 'numpy')
    f_diff_lambda = sp.lambdify(x, f_diff, 'numpy')

    p = p0
    for _ in range(max_iters):
        f_p = f_lambda(p)
        f_diff_p = f_diff_lambda(p)

        if abs(f_diff_p) < EPSILON:
            print('Derivative is too small. Aborting.')
            return None

        p_next = p - f_p / f_diff_p

        if abs(p_next - p) < tol:
            print('Anwer found!')
            return p_next

        p = p_next

    print("Max iterations reached!")
    return p


if __name__ == "__main__":
    f = x - sp.cos(x)
    root = newton_method(f, math.pi / 4)
    print("Root:", root)

File: newton-method/newton_method.m
function root = newton_method(f_expr, p0, max_iters, tol)
    % Define symbolic variable
    syms x

    % Convert string function expression to symbolic if needed
    if isa(f_expr, 'char') || isa(f_expr, 'string')
        f_sym = str2sym(f_expr);
    else
        f_sym = f_expr;
    end

    % Compute derivative
    f_diff = diff(f_sym, x);

    % Create function handles for numerical evaluation
    f_handle = matlabFunction(f_sym, 'Vars', x);
    f_diff_handle = matlabFunction(f_diff, 'Vars', x);

    % Initial guess
    p = p0;

    for i = 1:max_iters
        f_p = f_handle(p);
        f_diff_p = f_diff_handle(p);

        if abs(f_diff_p) < eps
            disp('Derivative is too small. Aborting.');
            root = NaN;
            return;
        end

        p_next = p - f_p / f_diff_p;

        if abs(p_next - p) < tol
            disp('Answer found!');
            root = p_next;
            return;
        end

        p = p_next;
    end

    disp('Max iterations reached!');
    root = p;
end